include/real/real.hpp

```diff
@@ -307,11 +308,18 @@ namespace boost {
                 }, _real_p->get_real_number());
             }
 
+            // this is used to control the amount of recursion that goes on in the distribution. Essentially, when we do a + b, we may 
+            // look at one level below them (if applicable, i.e., they're operations pointing to operands of their own). 
+
+            // we at most need to look at two levels.
+            // this is because in the case we have, say, x*a+x*b -> (a+b)*x, we may want to look at a and b's operands, if applicable, to see if
+            // they may be distributed as well. 
+            enum class RECURSION_LEVEL{ZERO, ONE, TWO};
+
             // a helper function for distributing when performing addition/subtraction 
             // the returned bool tells us whether we distributed or not, which is mostly useful when assign_and_return_void is true
             // since std::optional<real> would = std::null_opt if assign_and_return_void is false.
-
-            std::pair<bool, std::optional<real>> check_and_distribute(real & other, bool assign_and_return_void, OPERATION op) {
+            std::pair<bool, std::optional<real>> check_and_distribute(const real & other, const bool assign_and_return_void, const OPERATION op, const RECURSION_LEVEL rc_lvl) {
```

Comment:
this function must be well dcumented and examplained in the final documentation and for future students.